### ISSUE TITLE
Server closes connection on simple secondary index creation [data is on a shared virtualbox folder]

### ISSUE THREAD

**Issue by pdonorio** [rethinkdb/rethinkdb#3791]
Hello,
i have been running successfully a rethinkdb server on my docker cluster for months.
I designed some python API with flask that uses the server from another container.
I am starting to use indexes now following [these instructions](http://rethinkdb.com/docs/secondary-indexes/python).

I get errors if i run the first command
(table is 'data' and i is 'key', both existing and used in a query before this one)

```python
r.table(table).index_create(i).run()

  File "/usr/local/lib/python2.7/dist-packages/rethinkdb/ast.py", line 99, in run
    return c._start(self, **global_optargs)
  File "/usr/local/lib/python2.7/dist-packages/rethinkdb/net.py", line 295, in _start
    return self._send_query(query, global_optargs)
  File "/usr/local/lib/python2.7/dist-packages/rethinkdb/net.py", line 384, in _send_query
    response = self._read_response(query.token)
  File "/usr/local/lib/python2.7/dist-packages/rethinkdb/net.py", line 332, in _read_response
    response_header = self._sock_recvall(12)
  File "/usr/local/lib/python2.7/dist-packages/rethinkdb/net.py", line 258, in _sock_recvall
    raise RqlDriverError("Connection is closed.")
rethinkdb.errors.RqlDriverError: Connection is closed.
```

I just checked and i get the same error if i create the indexes via data explorer:

```javascript
r.db('webapp').table('data').indexCreate('key')
```

The database directory is placed in a shared directory on docker.
I am testing this in my development env (boot2docker on mac), so this means that the data directory in this case is a shared virtualbox directory.

I can try this same situation on a linux machine if necessary.

Paolo

**Comment by coffeemug** [comment]
@pdonorio -- sorry you ran into this.

Could you paste the output in the rethinkdb log (the log file in the RethinkDB data directory)?

**Comment by pdonorio** [comment]
Here the last piece of my rethinkdb log:

```bash
7d-b2e2-4c8a-bc85-b2be8c932253.create" (errno: 22 - Invalid argument). You may encounter data loss in case of a system failure. (Is the file located on a filesystem that doesn't support directory sync? e.g. VirtualBox shared folders)
2015-02-17T13:49:37.470950754 2192.579044s error: Error in src/arch/io/disk.cc at line 247:
2015-02-17T13:49:37.475854770 2192.583947s error: ftruncate failed.  (No such file or directory) (target size = 4194304)
2015-02-17T13:49:37.476654881 2192.584747s error: Backtrace:
2015-02-17T13:49:37.519940167 2192.628034s error: Tue Feb 17 13:49:37 2015\n\n1: backtrace_t::backtrace_t() at 0xb9d9f0 (/usr/bin/rethinkdb)\n2: format_backtrace(bool) at 0xb9dd83 (/usr/bin/rethinkdb)\n3: report_fatal_error(char const*, int, char const*, ...) at 0xd49fe5 (/usr/bin/rethinkdb)\n4: /usr/bin/rethinkdb() [0xac2050] at 0xac2050 ()\n5: linux_disk_manager_t::done(stats_diskmgr_t::action_t*) at 0xac49a3 (/usr/bin/rethinkdb)\n6: conflict_resolving_diskmgr_t::done(accounting_diskmgr_action_t*) at 0xab4594 (/usr/bin/rethinkdb)\n7: blocker_pool_t::on_event(int) at 0xaaf3e2 (/usr/bin/rethinkdb)\n8: epoll_event_queue_t::run() at 0xacfeee (/usr/bin/rethinkdb)\n9: linux_thread_pool_t::start_thread(void*) at 0xace6c5 (/usr/bin/rethinkdb)\n10: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8182) [0x7f6678969182] at 0x7f6678969182 (/lib/x86_64-linux-gnu/libpthread.so.0)\n11: clone+0x6d at 0x7f667869600d (/lib/x86_64-linux-gnu/libc.so.6)
2015-02-17T13:49:37.520234481 2192.628327s error: Exiting.
```

**Comment by coffeemug** [comment]
Ah, I see. It looks like this doesn't have anything to do with secondary indexes (this just happens to be the operation that triggers the error).

RethinkDB has an on-disk gabrage collector/compactor that incrementally removes old blocks from the file and truncates it. In some file system setups (e.g. virtualbox shared folders) these operations aren't supported, and the RethinkDB server exits with an error you got in the log file.

Is there a way to set up the data directory differently where it isn't in a shared VirtualBox folder?

**Comment by pdonorio** [comment]
I can verify this by running a test tomorrow on a linux host with docker, so i can bypass the virtualbox.
The problem is that I don't see how i could have data persistence if don't use the shared folder on boot2docker inside a docker container.

My suggestions:
- try to find a better way to show what the real problem is to a user in my same situation
- give an option to save indexes or other similar operations (unsupported on a shared virtualbox folder) into a separate directory. In such a way i could rebuild my indexes at each restart keeping persistence on tables data.

Thanks for your effort, and for your great work, by the way. I really love RethinkDB ;)

**Comment by coffeemug** [comment]
Cool, would be great if you could verify. Also, could paste the following:

- the first part of the log (I'd like to know which version of RethinkDB you're running)
- the output of `rethinkdb help`

We used to take advantage of DirectIO by default, but in newer versions we turned that off. Once I know the version you're running, I might be able to suggest a flag that might fix this.

Once we go through these steps, we'll formulate a more general solution to this problem and see if we can fix it in the product for everyone.

**Comment by pdonorio** [comment]
I just tested on linux and the problem is gone. So this is confirmed.
On linux i had rethinkdb 1.16.2+1~0trusty (GCC 4.8.2) and Linux 3.13.0-40-generic x86_64 

On my mac/boot2docker i am running instead
```bash
2015-02-05T12:48:47.993404134 0.090575s notice: Running rethinkdb 1.16.0+1~0trusty (GCC 4.8.2)...
2015-02-05T12:48:47.997444913 0.094618s notice: Running on Linux 3.16.7-tinycore64 x86_64
2015-02-05T12:48:47.997695288 0.094867s notice: Loading data from directory /data/rethinkdb_data
```

and:
```bash
root@b51f6c18f8b8:/rethinkdb# rethinkdb help
Running 'rethinkdb' will create a new data directory or use an existing one,
  and serve as a RethinkDB cluster node.
File path options:
  -d [ --directory ] path                     specify directory to store data and
                                              metadata
  --io-threads n                              how many simultaneous I/O operations
                                              can happen at the same time
  --direct-io                                 use direct I/O for file access
  --cache-size mb                             total cache size (in megabytes) for
                                              the process. Can be 'auto'.

Server name options:
  -n [ --server-name ] arg                    the name for this server (as will
                                              appear in the metadata).  If not
                                              specified, it will be randomly chosen
                                              from a short list of names.
  -t [ --server-tag ] arg                     a tag for this server. Can be
                                              specified multiple times.

Network options:
  --bind {all | addr}                         add the address of a local interface
                                              to listen on when accepting
                                              connections, loopback addresses are
                                              enabled by default
  --no-default-bind                           disable automatic listening on
                                              loopback addresses
  --cluster-port port                         port for receiving connections from
                                              other nodes
  --driver-port port                          port for rethinkdb protocol client
                                              drivers
  -o [ --port-offset ] offset                 all ports used locally will have this
                                              value added
  -j [ --join ] host[:port]                   host and port of a rethinkdb node to
                                              connect to
  --reql-http-proxy [protocol://]host[:port]  HTTP proxy to use for performing
                                              `r.http(...)` queries, default port
                                              is 1080
  --canonical-address addr                    address that other rethinkdb
                                              instances will use to connect to us,
                                              can be specified multiple times

Web options:
  --web-static-directory directory            the directory containing web
                                              resources for the http interface
  --http-port port                            port for web administration console
  --no-http-admin                             disable web administration console

CPU options:
  -c [ --cores ] n                            the number of cores to use

Service options:
  --pid-file path                             a file in which to write the process
                                              id when the process is running
  --daemon                                    daemonize this rethinkdb process

Set User/Group options:
  --runuser user                              run as the specified user
  --rungroup group                            run with the specified group

Help options:
  -h [ --help ]                               print this help
  -v [ --version ]                            print the version number of rethinkdb

Log options:
  --log-file file                             specify the file to log to, defaults
                                              to 'log_file'
  --no-update-check                           disable checking for available
                                              updates.  Also turns off anonymous
                                              usage data collection.

Configuration file options:
  --config-file                               take options from a configuration
                                              file


There are a number of subcommands for more specific tasks:
    'rethinkdb create': prepare files on disk for a new server instance
    'rethinkdb serve': use an existing data directory to host data and serve queries
    'rethinkdb proxy': serve queries from an existing cluster but don't host data
    'rethinkdb export': export data from an existing cluster into a file or directory
    'rethinkdb import': import data from from a file or directory into an existing cluster
    'rethinkdb dump': export and compress data from an existing cluster
    'rethinkdb restore': import compressed data into an existing cluster
    'rethinkdb index-rebuild': rebuild outdated secondary indexes

For more information, run 'rethinkdb help [subcommand]'.
```

**Comment by coffeemug** [comment]
Ok, I was going to suggest turning off direct IO, but in your version it's off by default.

One option here is to avoid exiting from the server, and instead print a warning into the log that `ftruncate` is failing. /cc'ing @danielmewes

**Comment by coffeemug** [comment]
Also, I'm not sure I'd classify this as a bug, as vbox shared folders legitimately limit the operation of the file system. I'd consider moving this to 2.0-polish, but leaving that up to @danielmewes.

**Comment by danielmewes** [comment]
Actually I think something slightly different is going on here, though this is still somewhat speculative.

First of all I don't think we can ignore `ftruncate` failures. We use `ftruncate` to increase the size of a data file among others. Even if a subsequent write to a high file offset would implicitly grow the file (which I'm not sure it would), we would still get read errors later because we assume that for example an extent is always completely allocated (and not just the parts we have actually been written to).

However I think the `ftruncate` is also not the problem here. `ftruncate` is a very very basic file operation. If that doesn't work, then I expect other operations to the file wouldn't work either. This is further supported by the error type that we get "No such file or directory".

I think what's actually going on here is something like this: When you create a secondary index, we briefly create a temporary file in `rethinkdb_data/tmp` that we use for storing a disk-backed queue. We open this file to get a file handle, and then immediately unlink it from the directory so it doesn't stay behind if the server is unexpectedly interrupted.
My suspicion is that VirtualBox shared folders don't follow the usual behavior of Unix file systems in this respect. I suspect that if you unlink a file on a shared folder, any file handles you still have on it get invalidated immediately. This would explain the "No such file or directory" error.

We'll have to test this, though if the theory is true I'm not sure if we can do much about it. We could skip unlinking the temporary file (risking leaking it, though rethinkdb would clean the tmp directory up on the next restart). We would need a way to detect whether the underlying file system supports our access pattern or not, to toggle between both cases.

I agree that this is not an super critical bug, more an incompatibility with VirtualBox shared folders.
It needs a bit of additional investigation. Once we've done that we can decide whether we want to
a) document the incompatibility, b) implement a work-around by detecting the file system and avoiding this usage pattern or c) file a bug with Oracle against VirtualBox.

**Comment by coffeemug** [comment]
:+1:

**Comment by pdonorio** [comment]
I vote for option 'c' ;)

I have a last question now: I tried to create my first index for the need of a orderBy with multiple fields,
from the [docs](http://www.rethinkdb.com/api/python/order_by/):

> You can efficiently order using multiple fields by using a compound index.

Do i have a workaround?

**Comment by danielmewes** [comment]
@pdonorio You can use `orderBy` with a function to get the same functionality without an index (it will be slower of course):
```
r.table(...).orderBy(function (x) { return [x('field1'), x('field2'), ...]; })
```

**Comment by danielmewes** [comment]
@alimoeeny also ran into this https://github.com/rethinkdb/rethinkdb/issues/3837 but in a boot2docker environment.

If this happens a lot, we should consider not unlinking the disk-backed queue files so early (assuming that's the problem). We clear out the tmp directory on restart anyway.

**Comment by coffeemug** [comment]
We could add a command line flag `--unlink-late` and recommend it to users that run into it (if it's hard to determine when to turn it on automatically).

**Comment by VeXocide** [comment]
As the two easiest ways of using Docker on a Mac, boot2docker and kitematic, are both based on VirtualBox I suspect we'll see this occurring more regularly and should find a way to solve it.

**Comment by danielmewes** [comment]
I think we should just never unlink the file while we're still using it. Instead we should only unlink it when the disk-backed queue is destroyed.

The only case where it makes a difference as far as I can tell is if the server crashes or is killed during a sindex construction or backfill. In that case the temporary file remains on disk until the server is restarted the next time. This doesn't seem like a huge problem to me.

**Comment by danielmewes** [comment]
...assuming that that is actually the problem.

**Comment by pdonorio** [comment]
> also ran into this #3837 but in a boot2docker environment.

Uhm, ok, just to make sure:
i am having the problem inside the boot2docker on a mac (which uses Virtualbox) too.

The shared folder is the only way to use data from my mac inside any container, which also leads to persistence of a data volume when the container is removed.

> You can use orderBy with a function to get the same functionality
> r.table(...).orderBy(function (x) { return [x('field1'), x('field2'), ...]; })

I have the python driver. Does the lambda function fit the case?

**Comment by danielmewes** [comment]
> Uhm, ok, just to make sure: i am having the problem inside the boot2docker on a mac (which uses Virtualbox) too.

Ah right, definitely the same issue then.

> I have the python driver. Does the lambda function fit the case?

Yes, you just have to write it slightly differently:
```py
r.table(...).orderBy(lambda x: [x['field1'], x['field2'], ...])
```

**Comment by coffeemug** [comment]
@danielmewes -- ah, I see. Makes complete sense to me.

**Comment by danielmewes** [comment]
The proposed change is up in CR 2664 by @AtnNn .

I tested this with VirtualBox and could reproduce the original issue, as well as confirm that this change fixes it.

**Comment by alimoeeny** [comment]
@danielmewes  You are fantastic. 
What does CR 2664 mean? Is there a dev branch that we can use? or is there a timeline for releases that I can look that up?

**Comment by danielmewes** [comment]
@alimoeeny The CR number is for our internal code review tool. You can find the fix in branch daniel_3791.

However it is branched off our development branch, so I wouldn't recommend using it for important data.

You could manually apply this patch https://github.com/rethinkdb/rethinkdb/commit/5e90224bf0485c40a8aa5f5812886f189e81f54f into the branch v1.16.x and then compile from there. Does that work for you?

The patch is officially going to ship with RethinkDB 2.0, which we hope to release about two weeks from now.

Let me know if one of those options works for you. If you need this urgently I could compile a special binary based on v1.16.x for you (which guest operating system are you using?).

**Comment by danielmewes** [comment]
The work-around is now in `next` https://github.com/rethinkdb/rethinkdb/commit/6ab9a7b224852ffd84f0680c7b5da3948d514c7c and we will ship it as part of RethinkDB 2.0.

**Comment by alimoeeny** [comment]
@danielmewes thank you very much. 
I appreciate your offer to compile a binary for me, but I think I'd better wait for the official release.

**Comment by stuartpb** [comment]
A user just encountered this in stuartpb/rethinkdb-dockerfiles#5 - glad to hear there's a fix in 2.0.

I also suggested an alternative solution in stuartpb/rethinkdb-dockerfiles#4.
